src/ui/components/Library/Team/View/NewTestRuns/TestRunsContextRoot.tsx

```diff
@@ -0,0 +1,168 @@
+import { useRouter } from "next/router";
+import {
+  Dispatch,
+  ReactNode,
+  SetStateAction,
+  createContext,
+  useDeferredValue,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+
+import { TestRun, getTestRunTitle } from "shared/test-suites/TestRun";
+import { useGetTeamRouteParams } from "ui/components/Library/Team/utils";
+import { useTestRuns } from "ui/components/Library/Team/View/NewTestRuns/hooks/useTestRuns";
+import { useSyncTestRunIdToUrl } from "ui/components/Library/Team/View/TestRuns/hooks/useSyncTestIdToUrl";
+
+type TestRunsContextType = {
+  filterByBranch: "all" | "primary";
+  filterByStatus: "all" | "failed";
+  filterByText: string;
+  filterByTextForDisplay: string;
+  selectTestRun: Dispatch<SetStateAction<string | null>>;
+  setFilterByBranch: Dispatch<SetStateAction<"all" | "primary">>;
+  setFilterByStatus: Dispatch<SetStateAction<"all" | "failed">>;
+  setFilterByText: Dispatch<SetStateAction<string>>;
+  testRuns: TestRun[];
+  testRunId: string | null;
+  testRunIdForDisplay: string | null;
+  spec: string | null;
+  setSpec: React.Dispatch<React.SetStateAction<string | null>>;
+};
+
+type TestRunsFilterContextType = {
+  filterByTime: "week" | "month";
+  startTime: Date;
+  endTime: Date;
+  setFilterByTime: Dispatch<SetStateAction<"week" | "month">>;
+};
+
+export const TestRunsFilterContext = createContext<TestRunsFilterContextType>(null as any);
+export const TestRunsContext = createContext<TestRunsContextType>(null as any);
+
+const daysAgo = (days: number) => {
+  const date = new Date();
+  date.setDate(date.getDate() - days);
+  // Set zero hours, so when used in Cache as key, it will be the same for the whole day.
+  date.setHours(0, 0, 0, 0);
+  return date;
+};
+
+export function TestRunsFilterContextRoot({ children }: { children: ReactNode }) {
+  const [filterByTime, setFilterByTime] = useState<"week" | "month">("week");
+
+  const startTime = useMemo(() => {
+    if (filterByTime === "week") {
+      return daysAgo(7);
+    }
+    if (filterByTime === "month") {
+      return daysAgo(30);
+    }
+    return daysAgo(7);
+  }, [filterByTime]);
+
+  const endTime = useMemo(() => {
+    return new Date();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [filterByTime]);
```

Comment:
I want to only update endTime (to Date.now()) if there's any change to `filterByTime`. I'm not sure of right pattern in react here.

Reply:
You could combine the two memos and return both `startTime` and `endTime`